File: combinations/combinations.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = []

        def dfs(A: List[int], B: List[int]):
            if len(A) == k:
                result.append(A)
                return

            for i in range(len(B)):
                if not A or A[-1] < B[i]:
                    dfs(A + [B[i]], B[:i] + B[i+1:])

        dfs([], [i for i in range(1, n+1)])

        return result

File: combinations/combinations.java
class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        doCombine(result, new ArrayList<Integer>(), 1, n, k);
        return result;
    }

    private void doCombine(List<List<Integer>> result, ArrayList<Integer> comb, int start, int n, int k) {
        if (k == 0) {
            result.add(new ArrayList<Integer>(comb));
            return;
        }

        for (int i = start; i <= n; i++) {
            comb.add(i);
            doCombine(result, comb, i+1, n, k-1);
            comb.remove(comb.size()-1);
        }
    }
}